// customPanel/PanelProperties.js

import * as idProps from './parts/idProps';
import * as nameProps from './parts/NameProps';
import * as responsibleProps from './parts/ResponsibleProps';
import * as guideProps from './parts/GuideProps';
import * as peopleProps from './parts/PeopleProps';
import * as rulesProps from './parts/RulesProps';
import * as serviceProps from './parts/ServiceProps';
import * as dataObjectProps from './parts/ServiceProps';


function createGeneralTabGroups(element) {

    var generalGroup = {
        id: 'general',
        label: 'اطلاعات عمومی',
        entries: []
    };

    idProps.getElement(generalGroup, element);
    nameProps.getElement(generalGroup, element);
    // procedureProps(generalGroup, element);

    var detailsGroup = {
        id: 'details',
        label: 'جزئیات',
        entries: []
    };
    // linkProps(detailsGroup, element);
    // eventProps(detailsGroup, element);

    var documentationGroup = {
        id: 'documentation',
        label: 'مستندات',
        entries: []
    };

    // documentationProps(documentationGroup, element);

    return [
        generalGroup,
        detailsGroup,
        documentationGroup
    ];

}

function createSecondaryTabGroups(element, bpmnFactory, elementRegistry, translate) {

    var secondaryGroup = {
        id: 'secondary',
        label: 'اطلاعات تکمیلی',
        entries: []
    };

    responsibleProps.getElement(secondaryGroup, element);
    peopleProps.getElement(secondaryGroup, element);
    rulesProps.getElement(secondaryGroup, element);
    guideProps.getElement(secondaryGroup, element);
    serviceProps.getElement(secondaryGroup, element);
    dataObjectProps.getElement(secondaryGroup, element);

    return [
        secondaryGroup
    ];

}


export function getTabs(element) {

    var generalTab = {
        id: 'general',
        label: 'اطلاعات عمومی',
        groups: createGeneralTabGroups(element)
    };

    if(element.type.indexOf('Task') != -1 || element.type == "bpmn:CallActivity" || element.type == "bpmn:SubProcess") {

        var secondaryTab = {
            id: 'secondary',
            label: 'اطلاعات تکمیلی',
            groups: createSecondaryTabGroups(element)
        };

        return [
            generalTab,
            secondaryTab
        ];
    }

    return [
        generalTab
    ];
}
